File: scran_multibatch_hvg.m
function [norm, hvg] = scran_multibatch_hvg(datafile, cellsub, batchvar, genes, normpars, hvgpars, options)
arguments
    datafile
    cellsub
    batchvar {string,char,cellstr}
    genes
    normpars.gene_subset=false
    normpars.do_multibatch=true
    normpars.do_pooledsizefactors=false
    normpars.min_mean=0.1 %key param
    hvgpars.n_features=0
    hvgpars.do_poissonvar=false
    hvgpars.do_densityweights=false %overrides fitTrendVar default of TRUE. important when HVGs are also high-abundance
    hvgpars.var_thr=0.0 %key param
    hvgpars.fdr_thr=1 % set to 1 to omit FDR threshold. overly conservative.
    hvgpars.min_mean_hvg=0.1 %key param
    options.tmp_path=[]
    options.tmp_fileroot="tmp"
    options.verbose=false
end
%TODO: support user clusters for pooled sfs
% gene subset? aka row_subset in scran?

% do_poissonvar = as.logical(hvgpars$do.poissonvar)
% do_topn = as.logical(hvgpars$do.topn)
% n_hvg = as.numeric(hvgpars$n.features)
% var_thr = as.numeric(hvgpars$var.thr)
% fdr_thr = as.numeric(hvgpars$fdr.thr)

if options.verbose
    disp("Running " + mfilename + "...")
end

if isempty(options.tmp_path)
    options.tmp_path="D:/tmp/tmp_"+mfilename+"/";
end

if ~exist(options.tmp_path,"dir")
    mkdir(options.tmp_path)
end

datafile=cellstr(datafile);

cellsubsetfile='D:/tmp/tmp_cellsub.csv';
writetable(cellsub,cellsubsetfile)
cellsubsetfile=cellstr(cellsubsetfile);

batchvar=cellstr(batchvar);

parfile=fullfile(options.tmp_path, options.tmp_fileroot+"_pars.mat");
save(parfile,"datafile","cellsubsetfile","batchvar","normpars","hvgpars")

Rpath = [getExternalPath("R"), filesep, 'Rscript.exe', '" "', '--vanilla '];
scriptfile = mfilename('fullpath')+".R";
command = char(strjoin([scriptfile,options.tmp_path, options.tmp_fileroot]," "));
commandline=['"', Rpath, command '"'];

tic
[status, cmdout]=system(commandline);

if status~=0
    error(cmdout)
end
if options.verbose
    disp(cmdout)
end
disp(mfilename+": ")

%TODO: link actual methods - pooledSFS vs libsize (pre multibatch)
normpars.method="scran_multibatchnorm";
hvgpars.method="scran_hvgs";

norm=normpars;
hvg=hvgpars;

%TODO: return regular libsizefactors
sfsfile=fullfile(options.tmp_path, options.tmp_fileroot+"_sfs.txt");
norm.sizefactors=readmatrix(sfsfile);
% norm.libsizefactors

%TODO: pass gene table in to get name/ix
hvgfile=fullfile(options.tmp_path, options.tmp_fileroot+"_hvgs.txt");
hvg.id=string(readcell(hvgfile));
hvg.ix= getGeneIndices(hvg.id,genes.id);
hvg.name = genes.name(hvg.ix);

meanfile=fullfile(options.tmp_path, options.tmp_fileroot+"_mean.txt");
varfile=fullfile(options.tmp_path, options.tmp_fileroot+"_var.txt");
biofile=fullfile(options.tmp_path, options.tmp_fileroot+"_bio.txt");
techfile=fullfile(options.tmp_path, options.tmp_fileroot+"_tech.txt");
hvg.mean=readmatrix(meanfile);
hvg.var=readmatrix(varfile);
hvg.bio=readmatrix(biofile);
hvg.tech=readmatrix(techfile);

if normpars.do_pooledsizefactors
    clustfile=fullfile(options.tmp_path, options.tmp_fileroot+"_qclust.txt");
    norm.clust=string(readcell(clustfile));
end
toc

File: scran_multibatch_hvg.R
# command line script to run multibatchnorm, modelGeneVar, and getTopHGVs
# suppressPackageStartupMessages(library(DropletUtils))
suppressPackageStartupMessages(library(scran))
suppressPackageStartupMessages(library(batchelor))
suppressPackageStartupMessages(library(R.matlab))

tmp_path <- "D:/tmp/tmp_scran_multibatch_hvg/"
tmp_fileroot <- "tmp"
args <- commandArgs(TRUE)
tmp_path <- args[1]
tmp_fileroot <- args[2]

print('Loading data...')
start_time = Sys.time()

parsmatfile<-file.path(tmp_path,paste0(tmp_fileroot,"_pars.mat"))
mat <- readMat(parsmatfile)
datafile <- unlist(mat$datafile) #full path to h5 file
cellsubsetfile <- unlist(mat$cellsubsetfile) #two column file with full list of cells. Colnames = id, keep
batchvar <- unlist(mat$batchvar)

normpars=mat$normpars[,,1]
do_pooledsizefactors = as.logical(normpars$do.pooledsizefactors)
do_multibatch = as.logical(normpars$do.multibatch)
min_mean = as.numeric(normpars$min.mean)

hvgpars=mat$hvgpars[,,1]
do_poissonvar = as.logical(hvgpars$do.poissonvar)
do_densityweights = as.logical(hvgpars$do.densityweights)
min_mean_hvg = as.numeric(hvgpars$min.mean.hvg)

var_thr = as.numeric(hvgpars$var.thr)
n_hvg = as.numeric(hvgpars$n.features)
if (n_hvg==0) { n_hvg = NULL}
fdr_thr = as.numeric(hvgpars$fdr.thr)
if (fdr_thr==1) { fdr_thr = NULL}

cellinfo <-read.csv(cellsubsetfile, row.names = 'id', header = T)
cellinfo <- as(cellinfo, "DataFrame")

data <- Seurat::Read10X_h5(datafile, use.names = F)
sce <- SingleCellExperiment(assays=list(counts=data))
colData(sce) <- cellinfo
sce <- sce[,sce$keep==1]

gene_sub = as.logical(normpars$gene.subset)
if (length(gene_sub)==dim(sce)[1]) {
  sce <- sce[gene_sub==T,]
}
block<-sce@colData[,batchvar]


print(Sys.time() - start_time)
print(dim(sce))


clust.pin <- NULL
if (do_pooledsizefactors==T) {
  print('computeSumFactors...')
  start_time = Sys.time()
  
  clust.pin <- quickCluster(sce, block=block)
  sce <- computeSumFactors(sce, cluster=clust.pin, min.mean=min_mean)
  
  print(Sys.time() - start_time)
}


start_time = Sys.time()

if (do_multibatch==T) {
  print('multiBatchNorm...')
  sce <- multiBatchNorm(sce, batch=block, min.mean=min_mean)
} else
{
  print('logNormCounts...')
  sce <- logNormCounts(sce)
}

print(Sys.time() - start_time)


print('modelGeneVar + getTopHVGs...')
start_time = Sys.time()

MEAN = matrix(data=NA_real_,nrow=n_hvg,ncol=length(min_mean_hvg))
VAR = matrix(data=NA_real_,nrow=n_hvg,ncol=length(min_mean_hvg))
BIO = matrix(data=NA_real_,nrow=n_hvg,ncol=length(min_mean_hvg))
TECH = matrix(data=NA_real_,nrow=n_hvg,ncol=length(min_mean_hvg))
for (i in seq(length(min_mean_hvg))) {
mm <- min_mean_hvg[[i]]
if (do_poissonvar==T) {
  blk <- modelGeneVarByPoisson(sce, block=block, min.mean=min_mean)
} else
{
  blk <- modelGeneVar(sce, block=block, min.mean=mm, density.weights=do_densityweights)
}

chosen.hvgs <- getTopHVGs(blk, n=n_hvg, var.threshold = var_thr, fdr.threshold = fdr_thr)

chosen.stats = blk[chosen.hvgs,]
MEAN[,i] = chosen.stats$mean
VAR[,i] = chosen.stats$total
BIO[,i] = chosen.stats$bio
TECH[,i] = chosen.stats$tech
}

print(Sys.time() - start_time)


print('Writing results to CSV files...')
start_time = Sys.time()

hvgfile<-file.path(tmp_path,paste0(tmp_fileroot,"_hvgs.txt"))
sfsfile<-file.path(tmp_path,paste0(tmp_fileroot,"_sfs.txt"))

writeLines(chosen.hvgs, con = hvgfile)
write.table(sizeFactors(sce), file=sfsfile, sep=',', row.names = F, col.names = F)

meanfile<-file.path(tmp_path,paste0(tmp_fileroot,"_mean.txt"))
varfile<-file.path(tmp_path,paste0(tmp_fileroot,"_var.txt"))
biofile<-file.path(tmp_path,paste0(tmp_fileroot,"_bio.txt"))
techfile<-file.path(tmp_path,paste0(tmp_fileroot,"_tech.txt"))
write.table(MEAN, file=meanfile, sep=',', row.names = F, col.names = F)
write.table(VAR, file=varfile, sep=',', row.names = F, col.names = F)
write.table(BIO, file=biofile, sep=',', row.names = F, col.names = F)
write.table(TECH, file=techfile, sep=',', row.names = F, col.names = F)

if (do_pooledsizefactors==T) {
  clustfile<-file.path(tmp_path,paste0(tmp_fileroot,"_qclust.txt"))
  write.table(clust.pin, file=clustfile, sep=',', row.names = F, col.names = F)
}

print(Sys.time() - start_time)
